### PR TITLE
Use stacklevel=2 in warnings.warn calls to DeprecationWarning where makes sense.

### LINKED ISSUE #7463: DeprecationWarning raised by Pylint's own code should generally have stacklevel=2
### Bug description

Looks like recently Pylint has added many `DeprecationWarning` in preparation for 3.0 (great!).

However, the `warnings.warn` call doesn't specify a `stacklevel=` thus default to `1`. Example: https://github.com/PyCQA/pylint/blob/27925396c85ffc443243589a0c03e430de9d7fa5/pylint/config/arguments_provider.py#L74

This makes the warning show up at the `warnings.warn` itself, like:

```
pylint/config/arguments_provider.py:74: DeprecationWarning: The checker-specific config attribute has been deprecated. Please use 'linter.config' to access the global configuration object.
```

It doesn't tell us where this deprecated method is called, making custom tooling that's built on top of Pylint difficult to migrate when 3.0 comes.

Most of the `warnings.warn` should use a `stacklevel=2` instead.

### Configuration

_No response_

### Command used

```shell
N/A
```

### Pylint output

```shell
N/A
```

### Expected behavior

N/A

### Pylint version

```shell
Pylint at head
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by yilei**
With this change, relevant `DeprecationWarnings` are now raised with `stacklevel=2`, so they have the callsite attached in the message.

## Type of Changes

I'm not sure this fits anything below?

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

Closes #7463

### CODE REVIEW

File: doc/whatsnew/fragments/7463.other
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/7463.other:1]:
```suggestion
Relevant `DeprecationWarnings` are now raised with `stacklevel=2`, so they have the callsite attached in the message.

Closes #7463 
```

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, I'm going to wait for a review by @DanielNoord before merging :)

**Review by DanielNoord — DISMISSED** [review]
Thanks! I didn't know about this argument, learned something new today 😄

**Review by DanielNoord — DISMISSED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**
